### ISSUE TITLE
Add option to exclude the compilation of OQS_randombytes_system() when a custom algorithm is specified

### ISSUE THREAD

**Issue by mabarger** [open-quantum-safe/liboqs#1490]
Compiling `liboqs` for systems without a (full) standard library (e.g. for embedded systems) causes problems at the moment, as `liboqs` uses multiple library functions (e.g. `fopen`), which might not be available on these systems.

Some of these calls are located in the function(s) `OQS_randombytes_system()` in [rand.c](https://github.com/open-quantum-safe/liboqs/blob/main/src/common/rand/rand.c) and will also be compiled  when providing a custom randombytes algorithm with `OQS_randombytes_custom_algorithm()`. Unless I am missing something the compilation of `OQS_randombytes_system()` is not necessary when a custom algorithm is provided and removing them from the compilation process could ease future use and integration.

I propose adding a new configuration option, which will enable/disable compilation of these functions by wrapping them in a preprocessor macro similar to this:

```C
@@ -35,7 +35,11 @@ void OQS_randombytes_openssl(uint8_t *random_array, size_t bytes_to_read);
 // Use OpenSSL's RAND_bytes as the default PRNG
 static void (*oqs_randombytes_algorithm)(uint8_t *, size_t) = &OQS_randombytes_openssl;
 #else
+#if !defined(OQS_NO_STDLIB)
 static void (*oqs_randombytes_algorithm)(uint8_t *, size_t) = &OQS_randombytes_system;
+#else
+static void (*oqs_randombytes_algorithm)(uint8_t *, size_t) = NULL;
+#endif
 #endif
 OQS_API OQS_STATUS OQS_randombytes_switch_algorithm(const char *algorithm) {
        if (0 == strcasecmp(OQS_RAND_alg_system, algorithm)) {
@@ -64,6 +68,7 @@ OQS_API void OQS_randombytes(uint8_t *random_array, size_t bytes_to_read) {
        oqs_randombytes_algorithm(random_array, bytes_to_read);
 }
 
+#if !defined(OQS_NO_STDLIB)
 #if !defined(_WIN32)
 #if defined(OQS_HAVE_GETENTROPY)
 void OQS_randombytes_system(uint8_t *random_array, size_t bytes_to_read) {
@@ -120,6 +125,7 @@ void OQS_randombytes_system(uint8_t *random_array, size_t bytes_to_read) {
        CryptReleaseContext(hCryptProv, 0);
 }
 #endif
+#endif
 
 #ifdef OQS_USE_OPENSSL
 #define OQS_RAND_POLL_RETRY 3 // in case failure to get randomness is a temporary problem, allow some repeats
```

This could also be extended to other library calls to improve the usability of `liboqs` in different settings.

**Comment by baentsch** [comment]
This sounds like a good proposal. Would you be willing to contribute a PR for this? Best including a test case for a system with such limitations?

**Comment by mabarger** [comment]
Yes, I'd love to contribute a PR for this.

For the test cases I was wondering what the ideal approach would be, as some of the default test cases would fail or need to be modified as they don't use a custom algorithm (except with the option`OQS_ENABLE_TEST_CONSTANT_TIME`).

For the test itself, what do we want to test for this configuration? Whether the problematic functions are still being included by analyzing the resulting library or something else?

**Comment by SWilson4** [comment]
This seems to be largely resolved by #1621.
